internal/migrate/versions.go

```diff
@@ -0,0 +1,209 @@
+	Steps    []*Step
+}
+
+// Step describes a single step in migration.
+type Step struct {
+	Description string
+	Version     int // Versions should start at 1
+	Action      Action
+}
+
+// Action is something that needs to be done
+type Action interface {
+	Run(log *zap.Logger, db DB, tx *sql.Tx) error
+}
+
+// ValidTableName checks whether the specified table name is valid
+func (migration *Migration) ValidTableName() error {
+	matched, err := regexp.MatchString(`^[a-z_]+$`, migration.Table)
+	if !matched || err != nil {
+		return Error.New("invalid table name: %v", migration.Table)
+	}
+	return nil
+}
+
+// Run runs the migration steps
+func (migration *Migration) Run(log *zap.Logger, db DB) error {
+	err := migration.ValidTableName()
+	if err != nil {
+		return err
+	}
+
+	err = migration.createVersionTable(log, db)
+	if err != nil {
+		return Error.New("creating version table failed: %v", err)
+	}
+
+	version, err := migration.getLatestVersion(log, db)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	if version < 0 {
+		log.Info("Latest Version", zap.Int("version", version))
+	} else {
+		log.Info("No Version")
+	}
+
+	if version < 0 && migration.OnCreate != nil {
+		log := log.Named("init")
+		tx, err := db.Begin()
+		if err != nil {
+			return Error.Wrap(err)
+		}
+
+		err = migration.OnCreate.Run(log, db, tx)
+		if err != nil {
+			return Error.Wrap(errs.Combine(err, tx.Rollback()))
+		}
+
+		err = migration.addVersion(tx, db, 0)
+		if err != nil {
+			return Error.Wrap(errs.Combine(err, tx.Rollback()))
+		}
+
+		if err := tx.Commit(); err != nil {
+			return Error.Wrap(err)
+		}
+	}
+
+	for _, step := range migration.Steps {
+		if step.Version <= version {
+			continue
+		}
+
+		log := log.Named(strconv.Itoa(step.Version))
+		log.Info(step.Description)
+
+		tx, err := db.Begin()
+		if err != nil {
+			return Error.Wrap(err)
+		}
+
+		err = step.Action.Run(log, db, tx)
+		if err != nil {
+			return Error.Wrap(errs.Combine(err, tx.Rollback()))
+		}
+
+		err = migration.addVersion(tx, db, step.Version)
+		if err != nil {
+			return Error.Wrap(errs.Combine(err, tx.Rollback()))
+		}
+
+		if err := tx.Commit(); err != nil {
+			return Error.Wrap(err)
+		}
+	}
+
+	return nil
+}
+
+// createVersionTable creates a new version tabe
+func (migration *Migration) createVersionTable(log *zap.Logger, db DB) error {
+	tx, err := db.Begin()
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	_, err = tx.Exec(db.Rebind(`CREATE TABLE IF NOT EXISTS ` + migration.Table + ` (version int, commited_at text)`))
+	if err != nil {
+		return Error.Wrap(errs.Combine(err, tx.Rollback()))
+	}
+
+	return Error.Wrap(tx.Commit())
+}
+
+// getLatestVersion finds the latest version table
+func (migration *Migration) getLatestVersion(log *zap.Logger, db DB) (int, error) {
+	tx, err := db.Begin()
+	if err != nil {
+		return -1, Error.Wrap(err)
+	}
+
+	var version sql.NullInt64
+	err = tx.QueryRow(db.Rebind(`SELECT MAX(version) FROM ` + migration.Table)).Scan(&version)
```

Comment:
why do we need to wrap a single SELECT in a transaction?

Reply:
Because that way we can keep the database interface requirements simpler.